handler.go

```diff
@@ -10,79 +10,151 @@ import (
+	CheckTx(ctx Context, store state.SimpleDB, tx Tx) (CheckResult, error)
 }
 
 // CheckerFunc (like http.HandlerFunc) is a shortcut for making wrapers
-type CheckerFunc func(Context, state.SimpleDB, Tx) (Result, error)
+type CheckerFunc func(Context, state.SimpleDB, Tx) (CheckResult, error)
 
-func (c CheckerFunc) CheckTx(ctx Context, store state.SimpleDB, tx Tx) (Result, error) {
+func (c CheckerFunc) CheckTx(ctx Context, store state.SimpleDB, tx Tx) (CheckResult, error) {
 	return c(ctx, store, tx)
 }
 
+// Deliver performs the tx once it makes it in the block
 type Deliver interface {
-	DeliverTx(ctx Context, store state.SimpleDB, tx Tx) (Result, error)
+	DeliverTx(ctx Context, store state.SimpleDB, tx Tx) (DeliverResult, error)
 }
 
 // DeliverFunc (like http.HandlerFunc) is a shortcut for making wrapers
-type DeliverFunc func(Context, state.SimpleDB, Tx) (Result, error)
+type DeliverFunc func(Context, state.SimpleDB, Tx) (DeliverResult, error)
 
-func (c DeliverFunc) DeliverTx(ctx Context, store state.SimpleDB, tx Tx) (Result, error) {
+func (c DeliverFunc) DeliverTx(ctx Context, store state.SimpleDB, tx Tx) (DeliverResult, error) {
 	return c(ctx, store, tx)
 }
 
-type SetOptioner interface {
-	SetOption(l log.Logger, store state.SimpleDB, module, key, value string) (string, error)
+// InitStater sets state from the genesis file
+type InitStater interface {
+	InitState(l log.Logger, store state.SimpleDB, module, key, value string) (string, error)
 }
 
-// SetOptionFunc (like http.HandlerFunc) is a shortcut for making wrapers
-type SetOptionFunc func(log.Logger, state.SimpleDB, string, string, string) (string, error)
+// InitStateFunc (like http.HandlerFunc) is a shortcut for making wrapers
+type InitStateFunc func(log.Logger, state.SimpleDB, string, string, string) (string, error)
 
-func (c SetOptionFunc) SetOption(l log.Logger, store state.SimpleDB, module, key, value string) (string, error) {
+func (c InitStateFunc) InitState(l log.Logger, store state.SimpleDB, module, key, value string) (string, error) {
 	return c(l, store, module, key, value)
 }
 
-// Result captures any non-error abci result
+// InitValidater sets the initial validator set
+type InitValidater interface {
+	InitValidate(log log.Logger, store state.SimpleDB, vals []*abci.Validator)
+}
+
+// InitValidateFunc (like http.HandlerFunc) is a shortcut for making wrapers
+type InitValidateFunc func(log.Logger, state.SimpleDB, []*abci.Validator)
+
+func (c InitValidateFunc) InitValidate(l log.Logger, store state.SimpleDB, vals []*abci.Validator) {
+	c(l, store, vals)
+}
+
+//---------- results and some wrappers --------
+
+type Dataer interface {
+	GetData() data.Bytes
+}
+
+// CheckResult captures any non-error abci result
 // to make sure people use error for error cases
-type Result struct {
+type CheckResult struct {
 	Data data.Bytes
 	Log  string
+	// GasAllocated is the maximum units of work we allow this tx to perform
+	GasAllocated uint
```

Comment:
Sure, I'll make it uint64.  I think that is overkill, but why not.  Also, it will be serialized at some point, so we need to specify 32 or 64.